File: Sources/ExcalidrawModel/Enums.swift
import Foundation

/// Fill pattern for shape backgrounds (`packages/element/src/types.ts: FillStyle`).
public enum FillStyle: String, Codable, Sendable, CaseIterable {
    case hachure
    case crossHatch = "cross-hatch"
    case solid
    case zigzag
}

/// Stroke line style (`StrokeStyle`).
public enum StrokeStyle: String, Codable, Sendable, CaseIterable {
    case solid
    case dashed
    case dotted
}

/// Horizontal text alignment (`TEXT_ALIGN`).
public enum TextAlign: String, Codable, Sendable, CaseIterable {
    case left
    case center
    case right
}

/// Vertical text alignment (`VERTICAL_ALIGN`).
public enum VerticalAlign: String, Codable, Sendable, CaseIterable {
    case top
    case middle
    case bottom
}

/// Image persistence status (`ExcalidrawImageElement.status`).
public enum ImageStatus: String, Codable, Sendable, CaseIterable {
    case pending
    case saved
    case error
}

/// Arrowhead styles, including legacy values still present in older files
/// (`Arrowhead` + `ArrowheadLegacy`).
public enum Arrowhead: String, Codable, Sendable, CaseIterable {
    case arrow
    case bar
    case circle
    case circleOutline = "circle_outline"
    case triangle
    case triangleOutline = "triangle_outline"
    case diamond
    case diamondOutline = "diamond_outline"
    case cardinalityOne = "cardinality_one"
    case cardinalityMany = "cardinality_many"
    case cardinalityOneOrMany = "cardinality_one_or_many"
    case cardinalityExactlyOne = "cardinality_exactly_one"
    case cardinalityZeroOrOne = "cardinality_zero_or_one"
    case cardinalityZeroOrMany = "cardinality_zero_or_many"
    // Legacy
    case dot
    case crowfootOne = "crowfoot_one"
    case crowfootMany = "crowfoot_many"
    case crowfootOneOrMany = "crowfoot_one_or_many"
}

/// Binding containment mode (`BindMode`).
public enum BindMode: String, Codable, Sendable, CaseIterable {
    case inside
    case orbit
    case skip
}

/// Font family ids are stored as integers in the file. Stable known values from
/// upstream `FONT_FAMILY`; kept as an `Int` namespace (not an enum) so unknown /
/// future ids round-trip losslessly.
public enum FontFamily {
    public static let virgil = 1
    public static let helvetica = 2
    public static let cascadia = 3
    public static let excalifont = 5
    public static let nunito = 6
    public static let lilitaOne = 7
    public static let comicShanns = 8
    public static let liberationSans = 9
    public static let assistant = 10

    /// Default for new text (`DEFAULT_FONT_FAMILY` = Excalifont).
    public static let `default` = excalifont
}

/// Roundness types (`ROUNDNESS`). Stored as an integer in `roundness.type`.
public enum RoundnessType {
    public static let legacy = 1
    public static let proportionalRadius = 2
    public static let adaptiveRadius = 3
}

File: web/packages/excalidraw-svelte/src/model/enums.ts
// String-valued enums from the .excalidraw schema. (parity: Enums.swift)

/** Fill pattern for shape backgrounds (`FillStyle`). */
export type FillStyle = "hachure" | "cross-hatch" | "solid" | "zigzag";

/** Stroke line style (`StrokeStyle`). */
export type StrokeStyle = "solid" | "dashed" | "dotted";

/** Horizontal text alignment (`TEXT_ALIGN`). */
export type TextAlign = "left" | "center" | "right";

/** Vertical text alignment (`VERTICAL_ALIGN`). */
export type VerticalAlign = "top" | "middle" | "bottom";

/** Image persistence status (`ExcalidrawImageElement.status`). */
export type ImageStatus = "pending" | "saved" | "error";

/** Binding containment mode (`BindMode`). */
export type BindMode = "inside" | "orbit" | "skip";

/**
 * Arrowhead styles, including legacy values still present in older files. Stored
 * as the raw string so unknown/future ids round-trip losslessly.
 */
export type Arrowhead =
  | "arrow"
  | "bar"
  | "circle"
  | "circle_outline"
  | "triangle"
  | "triangle_outline"
  | "diamond"
  | "diamond_outline"
  | "cardinality_one"
  | "cardinality_many"
  | "cardinality_one_or_many"
  | "cardinality_exactly_one"
  | "cardinality_zero_or_one"
  | "cardinality_zero_or_many"
  | "dot"
  | "crowfoot_one"
  | "crowfoot_many"
  | "crowfoot_one_or_many";

/** The element/tool types supported by Excalidraw (`ElementType`). */
export type ElementType =
  | "selection"
  | "rectangle"
  | "diamond"
  | "ellipse"
  | "arrow"
  | "line"
  | "freedraw"
  | "text"
  | "image"
  | "frame"
  | "magicframe"
  | "embeddable"
  | "iframe";

/**
 * Font family ids are stored as integers in the file. Stable known values from
 * upstream `FONT_FAMILY`; kept open (number) so unknown ids round-trip.
 */
export const FontFamily = {
  virgil: 1,
  helvetica: 2,
  cascadia: 3,
  excalifont: 5,
  nunito: 6,
  lilitaOne: 7,
  comicShanns: 8,
  liberationSans: 9,
  assistant: 10,
  /** Default for new text (`DEFAULT_FONT_FAMILY` = Excalifont). */
  default: 5,
} as const;

/** Roundness types (`ROUNDNESS`). Stored as an integer in `roundness.type`. */
export const RoundnessType = {
  legacy: 1,
  proportionalRadius: 2,
  adaptiveRadius: 3,
} as const;

/** `.excalidraw` schema constants (`ExcalidrawSchema`). */
export const ExcalidrawSchema = {
  schemaVersion: 2,
  fileType: "excalidraw",
} as const;
